### ISSUE TITLE
Assertion failed: (OrigType == signature.getReducedType(origType)), function initSwiftType, file AbstractionPattern.h, line 517.

### ISSUE THREAD

**Issue by dnadoba** [apple/swift#63332]
**Description**
5.8 & main nightly (30. January 2023) compiler crashes compiling the following code:

```swift
protocol Transform<Input, Output> {
    associatedtype Input
    associatedtype Output
}

struct Map<Input, Output>: Transform {
    var transform: (Input) -> Output
}

struct Chain2<A: Transform, B: Transform>: Transform where A.Output == B.Input {
    typealias Input = A.Input
    typealias Output = B.Output
    var a: A
    var b: B
}

extension Transform {
    @inlinable public func sorted<Element>(
        _ areInIncreasingOrder: @escaping (Element, Element) -> Bool
    ) -> some Transform<Input, [Element]> where Output == [Element]{
        map {
            $0
        }
    }
}

extension Transform {
    func map<NewValue>(
        _ transform: @escaping (Output) -> NewValue
    ) -> some Transform<Input, NewValue> {
        chain(Map(transform: transform))
    }
}

extension Transform {
    func chain<B: Transform>(
        _ transform: B
    ) -> Chain2<Self, B> {
        Chain2(a: self, b: transform)
    }
}
```

```
davidnadoba@Davids-M1-Max-2 PlaygroundPackage % xcrun --toolchain org.swift.58202301301a swift build
Building for debugging...
error: compile command failed due to signal 6 (use -v to see invocation)
Assertion failed: (OrigType == signature.getReducedType(origType)), function initSwiftType, file AbstractionPattern.h, line 517.
Please submit a bug report (https://swift.org/contributing/#reporting-bugs) and include the crash backtrace.
Stack dump:
0.	Program arguments: /Library/Developer/Toolchains/swift-5.8-DEVELOPMENT-SNAPSHOT-2023-01-30-a.xctoolchain/usr/bin/swift-frontend -frontend -c -primary-file /Users/davidnadoba/Repositories/PlaygroundPackage/Sources/PlaygroundPackage/main.swift -emit-dependencies-path /Users/davidnadoba/Repositories/PlaygroundPackage/.build/arm64-apple-macosx/debug/PlaygroundPackage.build/main.d -emit-reference-dependencies-path /Users/davidnadoba/Repositories/PlaygroundPackage/.build/arm64-apple-macosx/debug/PlaygroundPackage.build/main.swiftdeps -target arm64-apple-macosx13.0 -Xllvm -aarch64-use-tbi -enable-objc-interop -sdk /Applications/Xcode-14.2.0.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.1.sdk -I /Users/davidnadoba/Repositories/PlaygroundPackage/.build/arm64-apple-macosx/debug -I /Applications/Xcode-14.2.0.app/Contents/Developer/Platforms/MacOSX.platform/Developer/usr/lib -F /Applications/Xcode-14.2.0.app/Contents/Developer/Platforms/MacOSX.platform/Developer/Library/Frameworks -color-diagnostics -enable-testing -g -module-cache-path /Users/davidnadoba/Repositories/PlaygroundPackage/.build/arm64-apple-macosx/debug/ModuleCache -swift-version 5 -Onone -D SWIFT_PACKAGE -D DEBUG -new-driver-path /Library/Developer/Toolchains/swift-5.8-DEVELOPMENT-SNAPSHOT-2023-01-30-a.xctoolchain/usr/bin/swift-driver -entry-point-function-name PlaygroundPackage_main -empty-abi-descriptor -resource-dir /Library/Developer/Toolchains/swift-5.8-DEVELOPMENT-SNAPSHOT-2023-01-30-a.xctoolchain/usr/lib/swift -enable-anonymous-context-mangled-names -module-name PlaygroundPackage -target-sdk-version 13.1 -target-sdk-name macosx13.1 -o /Users/davidnadoba/Repositories/PlaygroundPackage/.build/arm64-apple-macosx/debug/PlaygroundPackage.build/main.swift.o -index-store-path /Users/davidnadoba/Repositories/PlaygroundPackage/.build/arm64-apple-macosx/debug/index/store -index-system-modules
1.	Apple Swift version 5.8-dev (LLVM 44d4f9d4b49845f, Swift 548a77341ef4875)
2.	Compiling with the current language version
3.	While evaluating request ASTLoweringRequest(Lowering AST to SIL for file "/Users/davidnadoba/Repositories/PlaygroundPackage/Sources/PlaygroundPackage/main.swift")
4.	While silgen emitFunction SIL function "@$s17PlaygroundPackage9TransformPAAE6sortedyQrSbqd___qd__tcSayqd__G6OutputRtzlF".
 for 'sorted(_:)' (at /Users/davidnadoba/Repositories/PlaygroundPackage/Sources/PlaygroundPackage/main.swift:18:23)
5.	While verifying SIL function "@$s17PlaygroundPackage9TransformPAAE6sortedyQrSbqd___qd__tcSayqd__G6OutputRtzlF".
 for 'sorted(_:)' (at /Users/davidnadoba/Repositories/PlaygroundPackage/Sources/PlaygroundPackage/main.swift:18:23)
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  swift-frontend           0x00000001077dfc7c llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 56
1  swift-frontend           0x00000001077def00 llvm::sys::RunSignalHandlers() + 128
2  swift-frontend           0x00000001077e02bc SignalHandler(int) + 304
3  libsystem_platform.dylib 0x000000019938f2a4 _sigtramp + 56
4  libsystem_pthread.dylib  0x0000000199360cec pthread_kill + 288
5  libsystem_c.dylib        0x000000019929a2c8 abort + 180
6  libsystem_c.dylib        0x0000000199299620 err + 0
7  swift-frontend           0x00000001078b05c8 swift::Lowering::AbstractionPattern::initSwiftType(swift::CanGenericSignature, swift::CanType, swift::Lowering::AbstractionPattern::Kind) (.cold.4) + 0
8  swift-frontend           0x00000001033d1ca8 swift::Lowering::AbstractionPattern::initSwiftType(swift::CanGenericSignature, swift::CanType, swift::Lowering::AbstractionPattern::Kind) + 128
9  swift-frontend           0x0000000103ad6abc (anonymous namespace)::SILTypeSubstituter::visitType(swift::CanType) + 248
10 swift-frontend           0x0000000103ad663c swift::CanTypeVisitor<(anonymous namespace)::SILTypeSubstituter, swift::CanType>::visit(swift::CanType) + 1652
11 swift-frontend           0x0000000103aceee0 swift::SILType::subst(swift::Lowering::TypeConverter&, llvm::function_ref<swift::Type (swift::SubstitutableType*)>, llvm::function_ref<swift::ProtocolConformanceRef (swift::CanType, swift::Type, swift::ProtocolDecl*)>, swift::CanGenericSignature, bool) const + 336
12 swift-frontend           0x0000000103aa57e0 swift::GenericEnvironment::mapTypeIntoContext(swift::SILModule&, swift::SILType) const + 156
13 swift-frontend           0x0000000103bccd04 (anonymous namespace)::SILVerifier::verifyEntryBlock(swift::SILBasicBlock*)::'lambda'(char const*, swift::SILType)::operator()(char const*, swift::SILType) const + 48
14 swift-frontend           0x0000000103bccb38 (anonymous namespace)::SILVerifier::visitSILFunction(swift::SILFunction*) + 9456
15 swift-frontend           0x0000000103bc6800 swift::SILFunction::verify(bool) const + 76
16 swift-frontend           0x00000001033cd3a4 swift::Lowering::SILGenModule::postEmitFunction(swift::SILDeclRef, swift::SILFunction*) + 228
17 swift-frontend           0x00000001033ccb5c swift::Lowering::SILGenModule::emitFunctionDefinition(swift::SILDeclRef, swift::SILFunction*) + 6948
18 swift-frontend           0x00000001033ce260 emitOrDelayFunction(swift::Lowering::SILGenModule&, swift::SILDeclRef) + 288
19 swift-frontend           0x00000001033cb020 swift::Lowering::SILGenModule::emitFunction(swift::FuncDecl*) + 164
20 swift-frontend           0x00000001034924c4 SILGenExtension::visitFuncDecl(swift::FuncDecl*) + 160
21 swift-frontend           0x000000010348e1dc SILGenExtension::emitExtension(swift::ExtensionDecl*) + 84
22 swift-frontend           0x000000010348e17c swift::Lowering::SILGenModule::visitExtensionDecl(swift::ExtensionDecl*) + 24
23 swift-frontend           0x00000001033d03a8 swift::ASTLoweringRequest::evaluate(swift::Evaluator&, swift::ASTLoweringDescriptor) const + 960
24 swift-frontend           0x0000000103483678 swift::SimpleRequest<swift::ASTLoweringRequest, std::__1::unique_ptr<swift::SILModule, std::__1::default_delete<swift::SILModule> > (swift::ASTLoweringDescriptor), (swift::RequestFlags)9>::evaluateRequest(swift::ASTLoweringRequest const&, swift::Evaluator&) + 148
25 swift-frontend           0x00000001033d43d8 llvm::Expected<swift::ASTLoweringRequest::OutputType> swift::Evaluator::getResultUncached<swift::ASTLoweringRequest>(swift::ASTLoweringRequest const&) + 400
26 swift-frontend           0x00000001033d136c swift::performASTLowering(swift::FileUnit&, swift::Lowering::TypeConverter&, swift::SILOptions const&, swift::IRGenOptions const*) + 100
27 swift-frontend           0x0000000102f1dc18 swift::performCompileStepsPostSema(swift::CompilerInstance&, int&, swift::FrontendObserver*) + 532
28 swift-frontend           0x0000000102f2c914 withSemanticAnalysis(swift::CompilerInstance&, swift::FrontendObserver*, llvm::function_ref<bool (swift::CompilerInstance&)>, bool) + 160
29 swift-frontend           0x0000000102f1f934 swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 3204
30 swift-frontend           0x0000000102d8ffcc swift::mainEntry(int, char const**) + 3304
31 dyld                     0x0000000199037e50 start + 2544
```
**Steps to reproduce**
Compile the above code or the attached SwiftPM package:
[PlaygroundPackage.zip](https://github.com/apple/swift/files/10551066/PlaygroundPackage.zip)

**Expected behavior**
no crash, Swift 5.7.2 compiles it just fine.

**Environment**
Tested with

Apple Swift version 5.8-dev (LLVM 44d4f9d4b49845f, Swift 548a77341ef4875)
Target: arm64-apple-macosx13.0

Apple Swift version 5.9-dev (LLVM dac17a9b87acc8e, Swift 3fd895a13da48fc)
Target: arm64-apple-macosx13.0

**Comment by AnthonyLatsis** [comment]
Silly reduction:
```swift
protocol Transform {
  associatedtype Input
  associatedtype Output
}

struct MapToInt<Input>: Transform {
  typealias Output = Int
}

extension Transform {
  func foo<U>(_: U) -> some Transform where Output == U {
    bar()
  }
}

extension Transform {
  func bar() -> some Transform {
    MapToInt<Output>()
  }
}
```

**Comment by dnadoba** [comment]
This is fixed in 5.9 🥳
